src/server/dcp_server/models.py

```diff
@@ -60,12 +75,475 @@ def masks_to_outlines(self, mask):
+         
+
+        Returns:
+            tuple: A tuple containing the cellpose_mask and final_mask, representing the segmentation masks obtained from
+                the Cellpose model and the combined segmentation and classification mask, respectively.
+        """
+
+        # Obtain segmentation mask using Cellpose model
+
+        # Find objects in the cellpose_mask
+        locs = find_objects(cellpose_mask)
+
+        # Get patches and labels based on object centroids
+        patches, labels = self.get_centered_patches(input_image, cellpose_mask, int(1.5 * input_image.shape[0] // 5), noise_intensity=5)
+        
+        labels = torch.tensor(labels)
+        labels_fit = []
+        
+        final_mask = torch.zeros(cellpose_mask.shape)
+
+        with torch.no_grad():
+            for i, patch in enumerate(patches):
+                loc = locs[i]
+
+                # Prepare image patch for classification
+                img = torch.tensor(patch.astype(np.float32)).unsqueeze(0).unsqueeze(0) / 255
+                # img = img.mean(dim=1, keepdim=True)
+              
+                # Perform inference using model_classifier
+                logits = self.classifier(img) 
+        
+                _, predicted = torch.max(logits, 1)
+                labels_fit.append(predicted)
+
+                # Assign predicted class to corresponding location in final_mask
+                final_mask[loc] = predicted + 1
+
+        # Apply mask to final_mask, retaining only regions where cellpose_mask is greater than 0
+        final_mask = final_mask * ((cellpose_mask > 0).long())
+        
+        return final_mask
+
+    def find_max_patch_size(self, mask):
+
+        # Find objects in the mask
+        objects = find_objects(mask)
+
+        # Initialize variables to store the maximum patch size
+        max_patch_size = 0
+        max_patch_indices = None
+
+        # Iterate over the found objects
+        for obj in objects:
+            # Extract start and stop values from the slice object
+            slices = [s for s in obj]
+            start = [s.start for s in slices]
+            stop = [s.stop for s in slices]
+
+            # Calculate the size of the patch along each axis
+            patch_size = tuple(stop[i] - start[i] for i in range(len(start)))
+
+            # Calculate the total size (area) of the patch
+            total_size = 1
+            for size in patch_size:
+                total_size *= size
+
+            # Check if the current patch size is larger than the maximum
+            if total_size > max_patch_size:
+                max_patch_size = total_size
+                max_patch_indices = obj
+            
+            max_patch_size_edge = np.ceil(np.sqrt(max_patch_size))
+
+            return max_patch_size_edge
+
+    def pad_centered_padded_patch(self, x: np.ndarray, c, p, mask: np.ndarray=None, noise_intensity=None) -> np.ndarray:
```

Comment:
I rename this function to crop_centered_padded_patch to avoid confusion

Reply:
Thanks, it's better :)